### ISSUE TITLE
Distutils package not bundled when run from virtualenv.

### ISSUE THREAD

**Issue by codewarrior0** [pyinstaller/pyinstaller#1420]
Built exe produces this error:
```
[...]
  File "...\IPython\lib\inputhook.py", line 25, in <module>
    from distutils.version import LooseVersion as V
ImportError: No module named version
```

Added a functional test in 27260b0abd05c81aa00c5453df572c9e6bf7eadd, currently marked xfail:

```
py.test -k test_distutils --runxfail
```

**Comment by olemoign** [comment]
I second this, having the same problem with my project.

**Comment by codewarrior0** [comment]
Upgrading this to `@high`. Having things work right in a virtualenv is important, as a lot of build systems (mine personally) will use a virtualenv for the build environment to make sure there are no surprises if a package is upgraded to a newer version.

**Comment by codewarrior0** [comment]
I looked into this briefly, but only got this far: The distutils hook will change the path for the distutils package itself, which gets `distutils/__init__.py` included, but the change doesn't affect any of the modules contained in the package.

**Comment by matysek** [comment]
Does it mean that pyinstaller would have to change path for all distutils modules?

**Comment by codewarrior0** [comment]
I don't know. More likely, it's not just distutils, but also any package module where a hook uses `retarget` to change the path. I think the hook worked before with the old 2.1 code (my project used to build correctly in a virtualenv), but now the hook only gets run after all of the imports are graphed? I admit, I haven't paid attention to changes in how hooks are run...

**Comment by htgoebel** [comment]
I'm afraid, we have a more serious problem here, see https://github.com/pyinstaller/pyinstaller/issues/1322#issuecomment-134702072:
> I had a look at this. As far as I understand, `modulegraph` *first* collects all imports recursively and *then* PyInstaller applies the hooks. This is too late for any hook changing the path or retrageting :-(

**Comment by matysek** [comment]
> I'm afraid, we have a more serious problem here, see #1322 (comment):
> 
> I had a look at this. As far as I understand, modulegraph first collects all imports recursively and then PyInstaller applies the hooks. This is too late for any hook changing the path or retrageting :-(

@htgoebel Does this mean that import hooks should be handled at the same stage as graph hooks?

**Comment by leycec** [comment]
I'm having similar difficulties in my futile attempts to fix the suite of `hook-wx.lib.pubsub*` hooks, which also attempt to modify their package's `__path__`.

## Just Hook It

It _would_ probably be feasible to merge import and graph hooks together – especially as there only exists one official graph hook and custom graph hooks are unsupported. Meaning, we can completely refactor graph hooks without worrying about breaking user applications. That's nice.

That said, graph hooks are run immediately _before_ the hooked module is imported and hence added to the graph. To resolve this and similar issues, we probably want import hooks to be run immediately _after_ the hooked module is imported and hence added to the graph. (For obvious reasons: the `hook()` function defined by import hooks expects to modify the passed `FakeModule` object containing the existing graph node.)

My recommendation would be to modify `PyiModuleGraph.import_hook()` to run all import hooks specific to the currently imported module immediately _after_ importing such module. Since I appear to need this functionality, perhaps I'll go ahead and hack that up.

## Backward Compatibility: _It's a Good Thing_

I should also note that **we've broken backward compatibility** in a pretty huge way, which I didn't quite realize until recently.

In the `develop` branch, the `hook()` function for a hook typically modified attributes of the passed module object. In the `python3` branch, that's (effectively) prohibited, as such modifications are silently ignored. Which is probably why so many hooks in the `python3` branch are currently broken.

_Yeah_.

## How Did `develop` Do It?

Is anyone still familiar with the `develop` branch? Clearly, the `develop` branch handles hook-based modifications of a module's `__path__` attribute in a semi-intelligent manner.

**What is that?** Do we have any comparable mechanisms in the `python3` branch?

I note the existing `FakeModule.retarget()` method, but that appears to assume that the passed path is a `.py`-prefixed uncompiled module (e.g., a package's `__init__.py` module). How would one retarget a namespace package, which contains no `__init__.py` module?

This is what I'm running up against with the `wx.lib.pubsub.pubsub1` package, which (basically) is a namespace package whose parent `wx.lib.pubsub.__init__.py` module hacks `sys.path` in such a way that the `wx.lib.pubsub.pubsub1` package magically **becomes** the `wx.lib.pubsub` package. (_It's crazy._)

_Anyway._ I suppose I should just take a look at the `develop` branch. :hushed:

**Comment by codewarrior0** [comment]
>I suspect that modulegraph simply ignores all new imports in the passed module. Why? Because the distutils package node and all children of that node already exist in the graph, so modulegraph lazily reuses those nodes without doing anything.

>Instead, modulegraph should:

>Note that the distutils package node has changed.
>Remove all child nodes of that node.
>Rescan the distutils package node's code object for new imports.

Hmm. If the graph edges are import statements, then the submodules of `distutils` won't really be the children of the `distutils` node. They will be children of whatever module is doing the `from distutils import version` (in my case it was something in IPython). You may have to scan the entire graph for `distutils` submodules.

>Rename "graph hooks" to "preimport hooks." 
>Rename "import hooks" to "postimport hooks."

Perfect.

>Incidentally, we've broken backward compatibility in a pretty huge way.

Yeah. Two big differences:

- Import hooks used to run during graph construction, and now they run afterward.
- FakeModule isn't a perfect emulation of the old `Module` object from `imptracker`.

>Ultimately, the current implementation of the FakeModule class is problematic.

I think FakeModule is just a shim to get the old `imptracker`-based code to work with the new `modulegraph` core. Sadly, the old code includes all of our hooks plus any user-defined hooks floating around out there, so it's probably here to stay. It definitely deserves some tidying-up and possibly a rename if that's the case.

>Would someone else be willing to tackle this – preferably someone familiar with the develop branch and how hooks used to work? I'd love to help here, but my PyInstaller plate is pretty full.

I could take a crack at it. I used to send patches for the `develop` branch before I found out the `python3` branch was also going to support Python 2.7, and since then I've mostly been improving the bootloader and the Windows-specific stuff. I'm currently cleaning out some rotten shared-assembly code from 2010, but when I'm done with that I can give FakeModule some attention.

**Comment by codewarrior0** [comment]
Also, this is completely unrelated, but I just found out ModuleGraph can output the graph as a .dot file for GraphViz, so [I tried it out on PyInstaller's `main.py`](http://i.imgur.com/YKbEt6b.png)

....pretty, but not what I was hoping for.

**Comment by leycec** [comment]
> Also, this is completely unrelated, but I just found out ModuleGraph can output the graph as a .dot file for GraphViz, so I tried it out on PyInstaller's main.py

**Looks about right.** Thanks to fundamental weaknesses in traditional visualization algorithms (e.g., force-directed stress majorization), GraphViz is known to scale horribly. I find graphs with more than 30 or so nodes to reliably bring GraphViz and, to be fair, _most_ graph visualizers to their decrepit knees.

**Comment by leycec** [comment]
> It definitely deserves some tidying-up and possibly a rename if that's the case.

It _definitely_ deserves a rename. Ideas are cheap, but I have none for sale. In the absence of any bolder suggestions, both `HookModule` and `PyiModule` seem sensible.

It deserves a bit more than a tidy-up, however. At the very least, PyInstaller should raise fatal exceptions on _any_ hooks attempting to modify `FakeModule` attributes. The current behaviour of user-defined hooks silently failing in inexplicable ways that used to work is completely unacceptable. (Please append all prior statements with "IMHO". _What the heck do I know, anyway?_)

Unlike Ruby, Python lacks a `freeze()` built-in for converting mutable into immutable objects. Nonetheless, it's not actually _that_ hard to immutable-ize objects. I note two ActiveState recipes, either of which would work for our limited purposes:

* [A Protocol for Making Objects Immutable](https://code.activestate.com/recipes/577895-a-protocol-for-making-objects-immutable). 2011.
* [ freeze(), make any object immutable](https://code.activestate.com/recipes/576527-freeze-make-any-object-immutable/). 2008.

The former recipe closely aligns with [PEP 351](http://www.python.org/dev/peps/pep-0351) – which, sadly, was rejected. But the core idea is sound, fairly simple, and gets us nearly all the way to a working solution. In summary:

1. Either:
   1. **Plan A:**
      1. 1. We redeclare all `FakeModule` attributes defined purely for compatibility with the old `imptracker.Module` class to be immutable objects. Python already supplies immutable alternatives for _most_ core types (e.g., `frozenset`, `tuple`), so this might get us all the way.
   1. **Plan B:**
      1. We define a new `PyInstaller.utils.misc.freeze()` function, inspired by one of the above recipes.
      1. We `freeze()` all `FakeModule` attributes defined purely for compatibility with the old `imptracker.Module` class that are not already inherently frozen. This means all non-string attributes, which, of course, are already immutable.
1. For each such attribute, define one `@property`-decorated getter (but not setter) defined in the trivial way. This permits attributes to be accessed but _not_ redefined.

Of course, this still breaks backward compatibility. But at least it explicitly informs the user that it's doing so.

**Comment by leycec** [comment]
> If the graph edges are import statements...

Yup!

> ...then the submodules of distutils won't necessarily be the children of the distutils node. They could be children of whatever module is doing the from distutils import version (in my case it was something in IPython). You may have to scan the entire graph for distutils submodules.

_Yup._ Nailed it on the natty head. :haircut: 

This won't be the most efficient operation ever, but it _should_ be feasible. The existing `altgraph.GraphUtils.filter_stack()` function appears to do everything we need, so it'll probably just be a matter of repurposing that. Hurrah!

**Comment by codewarrior0** [comment]
Before we get too crazy about modifying the entire graph with `filter_stack`, we should think some more about getting hooks to run *during* graph construction instead of *after*. If it's even possible.

But first, let's look at `FakeModule` some more.

Here's the old docs for `def hook()`:

```
``def hook(mod):``

    .. Note::

      The need to use this should be rare. Instead, try to use the
      global names described above first. This will keep the hook's
      code simple.

    Defines a function that takes a ``Module`` object.
    It must return a ``Module`` object, possibly the same one
    unchanged, or a modified one.
    A ``Module`` object is an instance of the class
    ``PyInstaller.depend.modules.Module()`` which you can read.
    If defined, ``hook(mod)`` is called before |PyInstaller| processed
    the global names described above.

    This function is supported to handle cases like dynamic modification of a
    package's ``__path__`` variable.
    A static list of names won't suffice
    because the new entry on ``__path__`` may well require computation.
    See ``hook-win32com.py`` in the hooks folder for an example.
```

And here are the things our own hooks do with the `Module`:

`hook-_tkinter.py`: `mod.add_data()`
`hook-distutils.py`: `mod.retarget()`
`hook-gst._gst.py`: `mod.add_binary()`
`hook-PIL.image.py`: `mod.add_import =` (probably broken, `add_import` is a function) and `mod.del_import()`
`hook-PIL.SpiderImagePlugin.py`: `enumerate(mod.imports)` and `del mod.imports[]`
`hook-PyQt4.uic.port_v2.py`: `mod.__path__ =`
`hook-wx.lib.pubsub.core.py`: `mod.__path__[]` and `mod.__path__.append()`

(some of the hooks I didn't mention are doing one of the above)

And one hook in a test suite that does a lot:

`tests\old_suite\basic\hooks1\hook-pkg1.py`: `mod.retarget()`, `mod.__path__ =`, `mod.node.packagepath = `, `mod.__path__[]`, `mod.del_import()`

I don't know what any feral user-defined hooks are doing, but they probably expect to receive a [`PyInstaller.depend.module.Module`](https://github.com/pyinstaller/pyinstaller/blob/develop/PyInstaller/depend/modules.py#L26) and will act accordingly. Actually, `Module` looks pretty useless from here, so let's see what PyI 2.1 used to do with hook files in [`ImportTracker.doimport()`](https://github.com/pyinstaller/pyinstaller/blob/develop/PyInstaller/depend/imptracker.py#L275)

Ah, they can use the globals `datas`, `binaries`, `hiddenimports`, and they can return a different `Module` to change its name and/or path. That last bit is probably the only thing we still have to deal with from wild hooks (if they even do it at all).

`PyInstaller.depend.module` is currently *gone*, so if any of them try to create a new `Module` and return it, they have been broken for a long time. On the other hand, they might just change the `__name__` or `__file__` in an attempt to redirect it to another module.

I'm not sure if `mod.add_data`, `mod.add_binary`, and `mod.add_import` are preferred over using the globals `datas`, `binaries`, and `hiddenimports` (respectively) in the hook module itself. 

`mod.retarget` is broken because, as we've mentioned often enough, hooks are run after graph construction instead of during.

`mod.imports` is filled in by FakeModule, but changes done to it by `hook()` have no effect.

`mod.__path__` is filled in by FakeModule and again, changes done to it by `hook()` have no effect. This comment is present where it is filled in:

... and there is a big chunk of TODOs at the top of FakeModule that mentions pretty much everything we've discussed.

--

So, two real questions:

- Which is preferred: `mod.add_data`, `mod.add_binary`, and `mod.add_import`? Or, the globals `datas`, `binaries`, and `hiddenimports`? The latter is used by PyI 2.1 hooks.

- How to fix `mod.retarget()`: Run the hooks after graph construction and rebuild the graph? Or run the hooks during graph construction and allow `retarget` to affect how the rest of the graph is constructed?

**Comment by leycec** [comment]
**Excellent.**

![Excellent.](http://i.imgur.com/wzHvW6j.gif)

> `PyInstaller.depend.module` is currently gone, so if any of them try to create a new `Module` and return it, they have been broken for a long time.

That raises an explicit exception, so I'm totally fine with that. Let laziness reign! :cocktail: 

> On the other hand, they might just change the `__name__` or `__file__` in an attempt to redirect it to another module.
> `mod.imports` is filled in by `FakeModule`, but changes done to it by `hook()` have no effect.
> `mod.__path__` is filled in by `FakeModule` and again, changes done to it by `hook()` have no effect. 

Those all silently fail. I'm less sanguine about that. Explicitly guarding against that probably took more time [to describe](https://github.com/pyinstaller/pyinstaller/issues/1420#issuecomment-135888136) than it would to actually implement, so... _yeah_.

I should probably just get on with that.

> Which is preferred: `mod.add_data`, `mod.add_binary`, and `mod.add_import`? Or, the globals `datas`, `binaries`, and `hiddenimports`? The latter is used by PyI 2.1 hooks.

Insert [Why Not Both?](https://www.youtube.com/watch?v=vqgSO8_cRio) memegirl here.

Ideally, we'd just murder the `mod.add_data()`, `mod.add_binary()`, and `mod.add_import()` methods. They duplicate the existing `datas`, `binaries`, and `hiddenimports` globals without tangible benefits – except that modifying globals in function scope requires use of the `global` declarator.

Is that a problem? Murphy's Law says, "Do seagulls greedily peck at your eyes?" Users _will_ forget to explicitly declare globals in function scope, resulting in their changes being (again) silently ignored: e.g.,

```
hiddenimports = [`muh.first.module`]

# It's all a bit awkward.
def hook(mod):
    # No "global hiddenimports" here means no "muh.other.module".
    hiddenimports.append(`muh.other.module`)
```

The `mod.add_data()`, `mod.add_binary()`, and `mod.add_import()` methods "just work." Users don't need to explicitly declare special magic to make that happen. So, I'm a reluctant fan of keeping both.

Bit awkward. _What can you do?_

> How to fix `mod.retarget()`: Run the hooks after graph construction and rebuild the graph?

**Woops.** That's a bit of a false dichotomy, but it's all my miserable fault. The `altgraph.GraphUtils.filter_stack()` function is silly and we should _never_ call it. Especially here.

No graph rebuild is required, thanks to the mystical wonders of `ModuleGraph.graph.nodes` – the  dictionary whose keys are fully-qualified module names and whose values are `Node` instances. Node removal efficiently reduces to:

1. `module_names = sorted(mod_graph.graph.nodes.keys())`, the lexicographically sorted list of all fully-qualified module names in `O(n logn)` time.
1. `distutils_index = bisect.bisect_right(module_names, 'distutils')`, the index of the `distutils` node in the `module_names` list obtained via binary search in `O(logn)` time.
1. Remove nodes _after_ `distutils_index` whose module names start with `distutils.` in `O(j)` time for `j` the number of such nodes.
1. Rescan the new `distutils.__init__` module in `O(k)` time for `k` the number of modules transitively imported by that module.

So, `O(n logn)` time and no additional space.

> Or run the hooks during graph construction and allow` retarget` to affect how the rest of the graph is constructed?

Definitely a possibility. While our existing preimport hook (i.e., graph hook) facilities are insufficient to support that, a refactor here and there should be sufficient to get them up to speed.

Preimport hooks are currently passed the current graph – and that's it. Dynamically retargeting modules during graph construction would require that preimport hooks be passed an instance of a new `PreimportModule` (..._or something, something_) class instead. This class serves a similar purpose to that of `FakeModule` (...which might be renamed to `PostimportHook`, under this hypothetical scenario), facilitating communication between PyInstaller and the preimport hook. It contains the current graph _and_ the three parameters passed to the `PyiModuleGraph._import_module()` method, permitting preimport hooks to "intercept" and modify the current call to this method. This class might have public attributes resembling:

```
class PreimportModule(object):
    # The current module graph.
    graph = ModuleGraph()

    # The fully-qualified name of the current module to be imported: e.g.,
    module_name = "email.mime.text"

    # The basename of the current module to be imported: e.g.,
    module_basename = "text"

    # Graph node for this module's package or
    # `None` if this module is a top-level module: e.g.,
    module_package = graph.findNode('email.mime')
```

By modifying the desired `PreimportModule` attributes in-place **before `modulegraph` even attempts an import,** preimport hooks provide a flexible means of dynamically rewriting the import graph with no space or time costs.

My. Beautiful. Gods. That actually sounds absurdly amazing and may very well live up to the above hype. "Why are we not doing this already, people?" ("Because yur laziness, leycec!")

Given that, we'd probably want to move the `retarget()` method from `FakeModule` (where it has no hoping of doing anything good) into `PreimportHook`. Imagine, if you will, a hypothetical preimport `distutils` hook:

```
def hook(preimport_module):
    """
    This hook checks for the distutils hacks present when using the
    virtualenv package.
    """
    # We need to do this in a subprocess, because distutils hasn't actually been
    # imported in the current process yet.
    distutils_path = exec_statement("""
import distutils
print(getattr(distutils, 'distutils_path', '')
""")

    # Non-empty  means PyInstaller is running inside virtualenv.
    # Virtualenv overrides real distutils modules.
    if distutils_path:
        distutils_init_filename = os.path.join(distutils_path, '__init__.py')
        preimport_module.retarget(distutils_init_filename)
```

We'd still need a separate postimport `distutils` hook, of course. Presumably, preimport hooks can't actually declare the `binaries`, `data`, or `hiddenimports` attributes. Although, I suppose we could permit that and then handle such attributes equivalently in both preimport and postimport hooks?

**ANYWAY.** For simplicity, let's just assume they can't. So, the existing `hooks/hook-distutils.py` hook would be split into two `hooks/preimport/distutils.py` and `hooks/postimport/distutils.py` hooks: the former retargeting `distutils` and the latter declaring the `binaries`, `data`, or `hiddenimports` attributes.

And we're done, folks.

**Comment by codewarrior0** [comment]
> The mod.add_data(), mod.add_binary(), and mod.add_import() methods "just work." Users don't need to explicitly declare special magic to make that happen. So, I'm a reluctant fan of keeping both.

Yep, a bit later I slipped in an edit when I realized there could be valid uses for both.

> My. Beautiful. Gods. That actually sounds absurdly amazing and may very well live up to the above hype. "Why are we not doing this already, people?" ("Because yur laziness, leycec!")

I am completely onboard with using `PreimportModule` and `PostimportModule` to insulate hooks from the details of how PyInstaller actually works with modules. I also like this solution better because

> Rescan the new distutils.__init__ module in O(k) time for k the number of modules transitively imported by that module.

This doesn't consider the fact that `distutils.__init__` isn't actually importing the submodules of `distutils`. It's usually going to be some other module, and we'd end up running all across the graph rewiring the edges to the newly retargeted submodules. Which could end up to be `O(not_worth_considering)`.

(dinner time, back in a bit)

**Comment by matysek** [comment]
> Before we get too crazy about modifying the entire graph with filter_stack, we should think some more about getting hooks to run during graph construction instead of after. If it's even possible.

Could it be possible to have part of hook file that will run during graph construction and part that will run after it?

> How to fix mod.retarget(): Run the hooks after graph construction and rebuild the graph? Or run the hooks during graph construction and allow retarget to affect how the rest of the graph is constructed?

Is there any way how to reconstruct the grahp with nodes that are reachable only from top-level module and all dead branches are removed?
* I somehow implemented hook attribute `excludedimports` and now the implementation does this:
  * remove the modules and submodules  for items from `excludedimports` when no other module imports them
  * this creates dead branches in the graph
  * PyInstaller bundles only modules that are reachable from the top-level node, without dead branches
  * downside: hooks are processed even for modules from the dead branches

**Comment by tallforasmurf** [comment]
This thing has gotten way beyond my depth, but I just want to add this and then shut up: with FakeModule I was *intentionally* breaking the hooks API, to the extent of requiring hooks to use methods add_whatever() instead of directly diddling the Module's contents. The point of requiring use of the methods was so that PyInstaller could directly *know* when a hook tried to change something -- as opposed to inspecting the module contents after the hook returned, to discover what it might have changed, if anything. (Is that even possible?)

I assumed at the time that breaking the API was OK because it was Python 3, the user will be changing all sorts of things, why not a hook too? Now of course you've got the erstwhile Python3 branch supporting 2.7, so suddenly breaking the API is not so easy to forgive.

**Comment by leycec** [comment]
> Is there any way how to reconstruct the grahp with nodes that are reachable only from top-level module and all dead branches are removed?

**Yup!** That's what the incredibly useful `ObjectGraph.flatten()` method does, which defers to the underlying `Graph.iterdata()` method for most of its dirty work.

Reconstructing the entire graph _would_ solve this issue but at extreme space and time costs paid on every hook performing retargeting. Thanks to codewarrior0's dinner-time parley [several comments up](https://github.com/pyinstaller/pyinstaller/issues/1420#issuecomment-135950089), it's evident that the better solution is to hook an appropriate `ModuleGraph` method (e.g., `_find_module()`, `_import_module()`) _while_ the graph is being reconstructed rather than attempt to restructure the graph _after_ the fact.

I'm reasonably convinced that I can implement this over the next several days with a minimum of pain and hassle. If I run into any deep blockers, I'll infect this thread with more complaining. :fire: 

Until then, best of luck with the bootloader! [_So say we all._](https://www.youtube.com/watch?v=B3EzRAgjo_s)

**Comment by leycec** [comment]
> Now of course you've got the erstwhile Python3 branch supporting 2.7, so suddenly breaking the API is not so easy to forgive.

**It's all good.** This is a collective volunteer effort. You took the invaluable first step. :bow: 

Given how dissimilar `imptracker` and `modulegraph` are, explicitly breaking backward compatibility by raising fatal exceptions on old hook behaviour is probably the best we can do. Fortunately, that should be pretty trivial. As [detailed above](https://github.com/pyinstaller/pyinstaller/issues/1420#issuecomment-135888136), converting mutable `FakeModule` attributes into immutable attributes mostly satisfies these requirements.

I'd happily hack on that... if I weren't already preoccupied by the indomitable Fortress of Suffering commonly referred to as `modulegraph`.

**Comment by leycec** [comment]
Almost done. :walking: 

`module_graph._find_module()` wasn't _quite_ the right method to be hooked. Instead, I've defined a new type of hook: the **post-package creation hook**, called immediately _after_ a package node is created but _before_ that package's `__init__.py` module (if any) is imported. Unsurprisingly, this is the only safe time to modify a package's `__path__` attribute and obtain the desired result.

To do so, I've:

* Defined a new `ModuleGraph._create_package()` method split from the existing `ModuleGraph._load_package()` method.
* Defined a new `imphook.PostCreatePackage` class. It's the `FakeModule` of the post-package creation world.
* Defined a new `PyiModuleGraph._create_package()` method running all applicable post-package hooks _after_ calling the superclass method.
* Defined a new `PyInstaller.hooks.post_create_package` package containing such hooks.
* Moved the existing `PyInstaller.hooks.hook-distutils` import hook to `PyInstaller.hooks.post_create_package.hook-distutils`.
* Refactored the latter to directly modify the `packagepath` of the newly created package node.

**So far, so good.** Extensive testing required, however. Expect an intimidating number of commits in two to three days.

When I've finished molested the codebase, we should probably remove the `FakeModule.retarget()` method. It never worked. Now, it never needs to.
